core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -678,6 +1074,159 @@ final class Stream[+F[_], +O] private (private val free: FreeC[Algebra[Nothing,
           .echo
     }.stream
 
+  /**
+    * Nondeterministically merges a stream of streams (`outer`) in to a single stream,
+    * opening at most `maxOpen` streams at any point in time.
+    *
+    * The outer stream is evaluated and each resulting inner stream is run concurrently,
+    * up to `maxOpen` stream. Once this limit is reached, evaluation of the outer stream
+    * is paused until one or more inner streams finish evaluating.
+    *
+    * When the outer stream stops gracefully, all inner streams continue to run,
+    * resulting in a stream that will stop when all inner streams finish
+    * their evaluation.
+    *
+    * When the outer stream fails, evaluation of all inner streams is interrupted
+    * and the resulting stream will fail with same failure.
+    *
+    * When any of the inner streams fail, then the outer stream and all other inner
+    * streams are interrupted, resulting in stream that fails with the error of the
+    * stream that caused initial failure.
+    *
+    * Finalizers on each inner stream are run at the end of the inner stream,
+    * concurrently with other stream computations.
+    *
+    * Finalizers on the outer stream are run after all inner streams have been pulled
+    * from the outer stream but not before all inner streams terminate -- hence finalizers on the outer stream will run
+    * AFTER the LAST finalizer on the very last inner stream.
+    *
+    * Finalizers on the returned stream are run after the outer stream has finished
+    * and all open inner streams have finished.
+    *
+    * @param maxOpen    Maximum number of open inner streams at any time. Must be > 0.
+    */
+  def join[F2[_], O2](maxOpen: Int)(implicit ev: O <:< Stream[F2, O2],
+                                    ev2: F[_] <:< F2[_],
+                                    F2: Concurrent[F2]): Stream[F2, O2] = {
+    assert(maxOpen > 0, "maxOpen must be > 0, was: " + maxOpen)
```

Comment:
Not entirely clear about how this is working. is `ev2` taking `F[_]` and `F2[_]` existentially? What makes this op different from the rest, the bound on `O`?

Reply:
A couple of things here -- `F[_] <:< F2[_]` isn't existential but rather evidence that `F` is a subtype of `F2`. This is the typelevel equivalent of the `F2[x] >: F[x]` bound on the type parameter.

In this case though, if we put that constraint on `F2` directly, then scalac fails to prove that `O` is `Stream[F2, O2]`, due to limitation in inferencing / type checking AFAICT. So instead, we leave `F2` unbound, let scalac derive `F2` via `O <:< Stream[F2, O2]` constraint, and then check if `F2` is a super type of `F`.